src/utils.ts

```diff
@@ -25,17 +26,19 @@ const AsyncFunctionConstructor = (async () => {}).constructor;
 /**
  * an async function check method only consider runtime support async syntax
  */
-export const isAsyncFunction = async (task: Task) => {
-  if (task.fn.constructor === AsyncFunctionConstructor) {
+export const isAsyncFunction = (fn: Fn) => fn.constructor === AsyncFunctionConstructor;
+
+export const isAsyncTask = async (task: Task) => {
+  if (isAsyncFunction(task.fn)) {
     return true;
   }
   try {
     if (task.opts.beforeEach != null) {
       await task.opts.beforeEach.call(task);
     }
     const call = task.fn();
-    const result = call instanceof Promise;
-    if (result) {
+    const result = typeof call?.then === 'function';
```

Comment:
Maybe `result` is too generic. Actually should be renamed to `promiseLike`.

Maybe create a typeguard isPromiseLike

See 
https://github.com/fastify/avvio/blob/master/lib/is-promise-like.js

Reply:
done it